src/AutorestSwift/JsonRpc/Client.swift

```diff
@@ -0,0 +1,225 @@
+// --------------------------------------------------------------------------
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//
+// The MIT License (MIT)
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the ""Software""), to
+// deal in the Software without restriction, including without limitation the
+// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+// sell copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//
+// --------------------------------------------------------------------------
+
+// TODO: Update license. Reproduced from: https://github.com/apple/swift-nio-examples/tree/master/json-rpc/Sources/JsonRpc
+
+import Foundation
+import NIO
+
+public final class ChannelClient {
+    private let group: MultiThreadedEventLoopGroup
+    public let config: Config
+    private var context: ChannelHandlerContext?
+    private let processCallback: ProcessCallback
+    private let logger = FileLogger(withFileName: "autorest-swift-debug.log")
+    // TODO: get this value based on last response id from AutoRest
+    var id = 3
+
+    public init(
+        group: MultiThreadedEventLoopGroup,
+        config: Config = Config(),
+        processCallback: @escaping ProcessCallback
+    ) {
+        self.group = group
+        self.config = config
+        self.processCallback = processCallback
+        self.state = .initializing
+    }
+
+    deinit {
+        assert(self.state == .stopped)
+    }
+
+    public func start(context: ChannelHandlerContext) {
+        assert(state == .initializing)
+
+        let future = context.pipeline.removeHandler(name: "AutorestServer").flatMap {
+            context.pipeline.removeHandler(name: "ServerHandler").flatMap {
+                context.pipeline.addHandler(
+                    CodableCodec<JSONResponse, JSONRequest>(),
+                    name: "AutorestClient"
+                ).flatMap {
+                    context.pipeline.addHandler(
+                        Handler(self.initalizationComplete),
+                        name: "ClientHandler"
+                    )
+                }
+            }
+        }
```

Comment:
Why do we need to remove the server handler? If it has both the client and server handlers simultaneously, what happens?

Reply:
If we leave both handlers simultaneously, the server handler will crash when we are  operating as a client. And vice versa. For exampole, in Server mode, we have this handler`CodableCodec<JSONRequest, JSONResponse >()`, it expects input as 'JSONResponse' which is not what it expects 'JSONRequest', it will cause casting error in handling the message.